ParProcCo/msmapper_aggregator.py

```diff
@@ -42,18 +52,137 @@ def _check_total_slices(self, total_slices: int, output_data_files: List[Path])
+
+        if "signal" in nxdata.attrs:
+            self.signal_name = nxdata.attrs["signal"]
+            if "axes" in nxdata.attrs:
+                self.axes_names = nxdata.attrs["axes"]
+                # TODO: add logic to create axes if not in file (from integer sequence starting at 0)
+                return
+        raise KeyError
+
+    def _fill_axes_fields(self) -> None:
+        self.axes_mins = [np.nan] * self.data_dimensions
+        self.axes_maxs = [np.nan] * self.data_dimensions
+        self._axes_starts_stops = [[] for _ in range(self.data_dimensions)]
+
+        for data_file in self.output_data_files:
+            with h5py.File(data_file, "r") as f:
+                # TODO: check default dataset is the same in all files
+                axes = [np.array(f[self.nxentry_name][self.nxdata_name][axis_name]) for axis_name in self.axes_names]
+            for j, axis in enumerate(axes):
+                self.axes_mins[j] = np.nanmin([np.nanmin(axis), self.axes_mins[j]])
+                self.axes_maxs[j] = np.nanmax([np.nanmax(axis), self.axes_maxs[j]])
+                self._axes_starts_stops[j].append([axis[0], axis[-1]])
+
+    def _initialise_accumulator_arrays(self) -> None:
+        self.accumulator_axis_lengths = []
+        self.accumulator_axis_ranges = []
+
+        for i in range(self.data_dimensions):
+
+            length = int((self.axes_maxs[i] - self.axes_mins[i]) / self.axes_spacing[i]) + 1
+            self.accumulator_axis_lengths.append(length)
+            ranges = [(round((x * self.axes_spacing[i]), 4) + self.axes_mins[i])
+                      for x in range(self.accumulator_axis_lengths[i])]
+            self.accumulator_axis_ranges.append(ranges)
+
+        self.accumulator_volume = np.zeros(self.accumulator_axis_lengths)
+        self.accumulator_weights = np.zeros(self.accumulator_axis_lengths)
+
+    def _accumulate_volumes(self) -> None:
+        for data_file in self.output_data_files:
+            with h5py.File(data_file, "r") as f:
+                # volumes and weights are 3d arrays
+                volumes_array = np.array(f[self.nxentry_name][self.nxdata_name][self.signal_name])
+                weights_array = np.array(f[self.nxentry_name][self.nxdata_name][self.aux_signal_name])
+                axes = [np.array(f[self.nxentry_name][self.nxdata_name][axis_name]) for axis_name in self.axes_names]
+
+            axes_lengths = volumes_array.shape
+            starts, stops = self._get_starts_and_stops(axes, axes_lengths)
+
+            if self.renormalisation:
+                volumes_array = np.multiply(volumes_array, weights_array)
+
+            self.accumulator_volume[starts[0]:stops[0], starts[1]:stops[1], starts[2]:stops[2]] += volumes_array
+            self.accumulator_weights[starts[0]:stops[0], starts[1]:stops[1], starts[2]:stops[2]] += weights_array
+
+    def _get_starts_and_stops(self, axes: List, axes_lengths: List) -> Tuple[List[int], List[int]]:
```

Comment:
Can combine this with `_fill_axes_fields` and `_initialise_accumulator_arrays` to check signal, auxiliary signals and axes exist in all input files as a pre-process and initialisation method.